File: flink-utils/src/main/java/io/github/ppdzm/utils/flink/deserialization/KafkaConsumerRecordDeserializationSchema.java
package io.github.ppdzm.utils.flink.deserialization;

import org.apache.flink.api.common.typeinfo.TypeHint;
import org.apache.flink.api.common.typeinfo.TypeInformation;
import org.apache.flink.streaming.connectors.kafka.KafkaDeserializationSchema;
import org.apache.kafka.clients.consumer.ConsumerRecord;

/**
 * @author Created by Stuart Alex on 2021/6/15.
 */
public class KafkaConsumerRecordDeserializationSchema implements KafkaDeserializationSchema<ConsumerRecord<String, String>> {
    private static final long serialVersionUID = -5095416451110247795L;

    @Override
    public boolean isEndOfStream(ConsumerRecord<String, String> stringStringConsumerRecord) {
        return false;
    }

    @Override
    public ConsumerRecord<String, String> deserialize(ConsumerRecord<byte[], byte[]> record) {
        String key = "", value = "";
        if (record.key() != null) {
            key = new String(record.key());
        }
        if (record.value() != null) {
            value = new String(record.value());
        }
        return new ConsumerRecord<>(
            record.topic(),
            record.partition(),
            record.offset(),
            record.timestamp(),
            record.timestampType(),
            record.checksum(),
            record.serializedKeySize(),
            record.serializedValueSize(),
            key,
            value
        );
    }

    @Override
    public TypeInformation<ConsumerRecord<String, String>> getProducedType() {
        return TypeInformation.of(new TypeHint<ConsumerRecord<String, String>>() {
        });
    }
}

File: flink-utils/src/main/scala/io/github/ppdzm/utils/flink/scala/deserialization/KafkaConsumerRecordDeserializationSchema.scala
package io.github.ppdzm.utils.flink.scala.deserialization

import org.apache.flink.api.common.typeinfo.TypeInformation
import org.apache.flink.streaming.connectors.kafka.KafkaDeserializationSchema
import org.apache.kafka.clients.consumer.ConsumerRecord

import java.nio.charset.StandardCharsets


/**
 * @author Created by Stuart Alex on 2022/4/6.
 */
class KafkaConsumerRecordDeserializationSchema extends KafkaDeserializationSchema[ConsumerRecord[String, String]] {
    override def isEndOfStream(t: ConsumerRecord[String, String]): Boolean = {
        false
    }

    override def deserialize(consumerRecord: ConsumerRecord[Array[Byte], Array[Byte]]): ConsumerRecord[String, String] = {
        val key =
            if (consumerRecord.key() != null) {
                new String(consumerRecord.key(), StandardCharsets.UTF_8)
            } else {
                null
            }
        val value =
            if (consumerRecord.value() != null) {
                new String(consumerRecord.value(), StandardCharsets.UTF_8);
            } else {
                null
            }
        new ConsumerRecord[String, String](
            consumerRecord.topic(),
            consumerRecord.partition(),
            consumerRecord.offset(),
            key,
            value
        )
    }

    override def getProducedType: TypeInformation[ConsumerRecord[String, String]] = {
        TypeInformation.of(classOf[ConsumerRecord[String, String]])
    }
}
